### ISSUE TITLE
`after_create_callback` is called before the transaction is commited started with `joinable: false`

### ISSUE THREAD

**Issue by kzkn** [rails/rails#46182]
### Steps to reproduce
```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "7.0.4"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :tests, force: true do |t|
    t.string :field
  end
end

class Test < ActiveRecord::Base
  attr_reader :called

  def initialize(attr = {})
    @called = false
    super
  end

  after_create_commit -> { @called = true }
end

class BugTest < Minitest::Test
  def test_with_joinable
    t = Test.new
    ActiveRecord::Base.transaction(joinable: true) do
      t.save
      raise ActiveRecord::Rollback
    end

    assert_equal false, t.called # => ok
  end

  def test_without_joinable
    t = Test.new
    ActiveRecord::Base.transaction(joinable: false) do
      t.save
      raise ActiveRecord::Rollback
    end

    assert_equal false, t.called # => failure
  end
end
```

### Expected behavior
The callback is not called in each case.

### Actual behavior
The callback is called in `test_without_joinable` case.

```
D, [2022-10-03T16:39:25.941413 #127693] DEBUG -- :   TRANSACTION (0.0ms)  begin transaction
D, [2022-10-03T16:39:25.941592 #127693] DEBUG -- :   TRANSACTION (0.1ms)  SAVEPOINT active_record_1
D, [2022-10-03T16:39:25.941816 #127693] DEBUG -- :   Test Create (0.1ms)  INSERT INTO "tests" DEFAULT VALUES
D, [2022-10-03T16:39:25.942085 #127693] DEBUG -- :   TRANSACTION (0.1ms)  RELEASE SAVEPOINT active_record_1
D, [2022-10-03T16:39:25.942351 #127693] DEBUG -- :   TRANSACTION (0.1ms)  rollback transaction
```

### System configuration
**Rails version**:
7.0.4

**Ruby version**:
ruby 3.0.4p208 (2022-04-12 revision 3fa771dded) [x86_64-linux]

**Comment by byroot** [comment]
As far as I know `joinable` is private API for internal use (transactional fixtures).

Could you explain why you think this should be the expected behavior, and why you are using `joinable: false`?

**Comment by kzkn** [comment]
Although the test code does not nest transactions, I often use the `joinable: false` option for rollback of nested transactions.
For example, in a common process within an application, I might specify `joinable: false` for rollback of nested transactions, in which transactions may or may not be nested.

```ruby
def within_transaction
  # before process...
  result = ApplicationRecord.transaction(joinable: false) do
    yield
  end
  # after process...
  result
end

class User < ApplicationRecord
  has_many :posts, dependent: :destroy
  class << self
    def some_batch_process
      transaction do
        # blah, blah
      end
    end
  end
end

# nested case
def process1
  within_transaction do
    User.some_batch_process
  end
end

# not nested case
def process2(user)
  within_transaction do
    user.posts.where(created_at: ..1.year.ago).destroy_all
    user.update_column(last_cleaned_at: Time.current)
  end
end
```

The test code I showed does not explicitly nest transactions. It looks like I am not committing as far as I can follow the letter of the code. But internally, a `SAVEPOINT`/`RELEASE SAVEPOINT` is performed through the `save` call, and `after_commit` is unexpectedly called, which is confusing.
In such a case, I think `after_commit` callback should not be called.

---

Also, if the `joinable` option is a private API and should not be used in the application code, it would be better to make that readily available from the documentation.

https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/DatabaseStatements.html#method-i-transaction

> transaction(requires_new: nil, isolation: nil, joinable: true, &block) 

The current documentation certainly does not explain the `joinable` option, but it appears to be a usable method signature.

**Comment by casperisfine** [comment]
> The current documentation certainly does not explain the `joinable` option, but it appears to be a usable method signature.

AFAIK `rdoc` doesn't allow to hide some parameters. Parameters that are not explicitly documented should be considered private.

If you look at the usage of `joinable: false` in Rails, it's used by transactional fixtures and the sandbox console to create an "invisible" transaction, AKA a transaction for which all the AR code behave like it wasn't there, hence call `after_commit`.

I'll dig a bit in the history to double check wether this was meant to be publicly usable, but I'm 99% certain it's not the case.

**Comment by casperisfine** [comment]
So I confirm it was added specifically for transactional fixtures and not intended for public use: https://github.com/rails/rails/commit/ab0ce052ba23a4cce7a84ecade0d00d9cc518ebd

**Comment by byroot** [comment]
@kzkn I found a way to hide a parameter from RDoc, so I did just that: https://github.com/rails/rails/pull/46186

That said, I understand your use case, and wouldn't be opposed to make `joinable` a supported option, but that require a refactoring so that transactional fixtures and console sandbox use another option and stop relying on `joinable?`.

I may work on this if I have some time, but no guarantees. If you really care about this I encourage you to try to implement it as a PR, I'll happily review.
